geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/FeatureAggregatingIterator.scala

```diff
@@ -0,0 +1,147 @@
+
+import java.util.{Collection => JCollection, Map => JMap}
+
+import com.typesafe.scalalogging.slf4j.Logging
+import org.apache.accumulo.core.data.{ByteSequence, Key, Range => ARange, Value}
+import org.apache.accumulo.core.iterators.{IteratorEnvironment, SortedKeyValueIterator}
+import org.geotools.feature.simple.SimpleFeatureBuilder
+import org.geotools.geometry.jts.JTSFactoryFinder
+import org.locationtech.geomesa.core._
+import org.locationtech.geomesa.core.data.DEFAULT_ENCODING
+import org.locationtech.geomesa.core.iterators.FeatureAggregatingIterator.Result
+import org.locationtech.geomesa.feature.{AvroSimpleFeatureFactory, FeatureEncoding, SimpleFeatureDecoder, SimpleFeatureEncoder}
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
+import org.opengis.feature.simple.SimpleFeatureType
+
+import scala.util.Random
+
+abstract class FeatureAggregatingIterator[T <: Result](val other: FeatureAggregatingIterator[T],
+                                                       val env: IteratorEnvironment)
+  extends SortedKeyValueIterator[Key, Value] {
+
+  var curRange: ARange = null
+  var projectedSFT: SimpleFeatureType = null
+  var featureBuilder: SimpleFeatureBuilder = null
+  var topKey: Option[Key] = None
+  var topValue: Option[Value] = None
+
+  var simpleFeatureType: SimpleFeatureType = null
+  var source: SortedKeyValueIterator[Key,Value] = null
+
+  var originalDecoder: SimpleFeatureDecoder = null
+  var featureEncoder: SimpleFeatureEncoder = null
+
+  var projectedSFTDef: String = null
+
+  if (other != null && env != null) {
+    source = other.source.deepCopy(env)
+    simpleFeatureType = other.simpleFeatureType
+  }
+
+  def init(source: SortedKeyValueIterator[Key, Value],
+           options: JMap[String, String],
+           env: IteratorEnvironment): Unit = {
+    this.source = source
+
+    val simpleFeatureTypeSpec = options.get(GEOMESA_ITERATORS_SIMPLE_FEATURE_TYPE)
+    simpleFeatureType = SimpleFeatureTypes.createType(this.getClass.getCanonicalName, simpleFeatureTypeSpec)
+    simpleFeatureType.decodeUserData(options, GEOMESA_ITERATORS_SIMPLE_FEATURE_TYPE)
+
+    // default to text if not found for backwards compatibility
+    val encodingOpt = Option(options.get(FEATURE_ENCODING)).map(FeatureEncoding.withName).getOrElse(DEFAULT_ENCODING)
+    originalDecoder = SimpleFeatureDecoder(simpleFeatureType, encodingOpt)
+
+    initProjectedSFTDefClassSpecificVariables(source, options, env)
+
+    projectedSFT = SimpleFeatureTypes.createType(simpleFeatureType.getTypeName, projectedSFTDef)
+
+    // Use density SFT for the encoder since we are transforming the feature into
+    // a sparse matrix as the result type of this iterator
+    featureEncoder = SimpleFeatureEncoder(projectedSFT, encodingOpt)
+    featureBuilder = AvroSimpleFeatureFactory.featureBuilder(projectedSFT)
+  }
+
+  def initProjectedSFTDefClassSpecificVariables(source: SortedKeyValueIterator[Key, Value],
```

Comment:
Since this a function to be overridden, I'd love to see a comment explaining what's going on.